src/environmentd/tests/sql.rs

```diff
@@ -1291,6 +1291,71 @@ fn test_explain_timestamp_json() {
     let _explain: TimestampExplanation<Timestamp> = serde_json::from_str(&explain).unwrap();
 }
 
+// Verify that `EXPLAIN TIMESTAMP ...` within acts like a peek within a transaction.
+// That is, ensure the following:
+// 1. Consistently returns its transaction timestamp as the "query timestamp"
+// 2. Acquires read holds for all objects within the same time domain
+// 3. Errors during a write-only transaction
+// 4. Errors when an object outside the chosen time domain is referenced
+#[test]
+fn test_github_18950() {
+    let config = util::Config::default();
+    let server = util::start_server(config).unwrap();
+    let mut client = server.connect(postgres::NoTls).unwrap();
+    client.batch_execute("CREATE TABLE t1 (i1 int)").unwrap();
+
+    // Verify execution during a write-only trx fails
+    client.batch_execute("BEGIN").unwrap();
+    client.batch_execute("INSERT INTO t1 VALUES (1)").unwrap();
+    let error = client
+        .query_one("EXPLAIN TIMESTAMP FOR SELECT * FROM t1;", &[])
+        .unwrap_err();
+
+    assert!(format!("{}", error).contains("transaction in write-only mode"));
+
+    client.batch_execute("ROLLBACK").unwrap();
+
+    // Verify the transaction timestamp is returned for each select and
+    // read holds are acquired as of the transaction timestamp
+    client.batch_execute("BEGIN").unwrap();
+    let mut query_timestamp = None;
+
+    for i in 1..5 {
+        let row = client
+            .query_one("EXPLAIN TIMESTAMP AS JSON FOR SELECT * FROM t1;", &[])
+            .unwrap();
+
+        let explain: String = row.get(0);
+        let explain: TimestampExplanation<Timestamp> = serde_json::from_str(&explain).unwrap();
+        let explain_timestamp = explain.determination.timestamp_context.timestamp().unwrap();
+
+        if let Some(timestamp) = query_timestamp {
+            assert_eq!(timestamp, *explain_timestamp);
+        } else {
+            query_timestamp = Some(*explain_timestamp);
+        }
+
+        for source in &explain.sources {
+            for timestamp in &source.read_frontier {
+                assert_eq!(*timestamp, query_timestamp.unwrap());
+            }
+        }
```

Comment:
Ah, yeah. I previously incorrectly assumed that acquiring read holds as of T is guaranteed to work. However, it looks like in the case that T is beyond a source's `read_frontier`, we acquire a hold at that time instead. I hadn't considered this, but defaulting to acquiring the hold at the lowest possible time that the source supports makes sense to me.

Thanks for catching that! I'll try and update the test case accordingly

Reply:
Ah ok, yeah in general acquiring the read hold isn't guaranteed to work, though I think in this particular scenario we know that it will work because we chose the read hold time based on the read frontier's of the objects. On the other hand, if we do successfully acquire a read hold at time `t` on object `o`, then we cannot assert `o`'s read frontier is equal to `t`. We can only assert that the read frontier is less than or equal to `t`.

Reply:
> if we do successfully acquire a read hold at time t on object o, then we cannot assert o's read frontier is equal to t. We can only assert that the read frontier is less than or equal to t.

The first sentence here makes sense to me, but not the second. If `o` has already compacted past `t`, then we acquire the hold at `o`'s read frontier, right? That's what I gathered from: https://github.com/MaterializeInc/materialize/blob/fa6bb00645617ac4dc9ccbff26f9181ef1d3c5fa/src/adapter/src/coord/read_policy.rs#L389-L432

Definitely could be reading that incorrectly, though.

I've updated the test to check that the initially reported read frontier does not change on subsequent iterations. If my logic is right above, then this is the most meaningful assertion I can think of.

Reply:
>The first sentence here makes sense to me, but not the second.

Let's say that an object `o` has a read frontier of `t - 100`, then we a acquire a read hold on object `o` at time `t`. The read frontier of `o` is not forcibly compacted up to `t`. It may remain at `t - 100` or some value less than `t` for some time. We're only guaranteed that the read frontier will not compact past `t`.

> If o has already compacted past t, then we acquire the hold at o's read frontier, right?

This is also correct.

Reply:
Ah I see what you were saying now. Makes sense, thanks for clarifying!